### ISSUE TITLE
JSReport 1.0+/mongodb

### ISSUE THREAD

**Issue by jcfergus** [jsreport/jsreport#231]
Has mongodb-store been tested with JSReport 1.0?  I'm having a heck of a time trying to get my install upgraded.  I have my jsreport install running, but once I get to the startup screen, any time I click on one of my templates or scripts or whatever, I get: 

Cannot read property '_id' of undefined

That corresponds with the following:

```
client.js?5ac7899…:61 XHR finished loading: GET "http://[redacted].net/reporting/odata/templates(57717d9e387844200057f98a)".f.end @ client.js?5ac7899…:61(anonymous function) @ client.js?5ac7899…:28n._execute @ client.js?5ac7899…:26o._resolveFromExecutor @ client.js?5ac7899…:27o @ client.js?5ac7899…:27g.(anonymous function) @ client.js?5ac7899…:28(anonymous function) @ client.js?5ac7899…:93o @ client.js?5ac7899…:61(anonymous function) @ client.js?5ac7899…:61e.(anonymous function) @ client.js?5ac7899…:61r @ client.js?5ac7899…:2(anonymous function) @ client.js?5ac7899…:2t @ client.js?5ac7899…:1(anonymous function) @ client.js?5ac7899…:2(anonymous function) @ client.js?5ac7899…:93(anonymous function) @ client.js?5ac7899…:93o @ client.js?5ac7899…:61(anonymous function) @ client.js?5ac7899…:61e.(anonymous function) @ client.js?5ac7899…:61r @ client.js?5ac7899…:2(anonymous function) @ client.js?5ac7899…:2t @ client.js?5ac7899…:1(anonymous function) @ client.js?5ac7899…:2(anonymous function) @ client.js?5ac7899…:93(anonymous function) @ client.js?5ac7899…:91(anonymous function) @ client.js?5ac7899…:91onClick @ client.js?5ac7899…:93onClick @ client.js?5ac7899…:92r @ client.js?5ac7899…:59a @ client.js?5ac7899…:89s @ client.js?5ac7899…:89d @ client.js?5ac7899…:2f @ client.js?5ac7899…:2n @ client.js?5ac7899…:60processEventQueue @ client.js?5ac7899…:2r @ client.js?5ac7899…:90handleTopLevel @ client.js?5ac7899…:90a @ client.js?5ac7899…:90i @ client.js?5ac7899…:90perform @ client.js?5ac7899…:28batchedUpdates @ client.js?5ac7899…:59i @ client.js?5ac7899…:1dispatchEvent @ client.js?5ac7899…:90
client.js?5ac7899…:61 XHR finished loading: GET "http://[redacted].net/reporting/odata/templates?$top=5&$select=name,recipe,modificationDate&$orderby=modificationDate%20desc".f.end @ client.js?5ac7899…:61(anonymous function) @ client.js?5ac7899…:28n._execute @ client.js?5ac7899…:26o._resolveFromExecutor @ client.js?5ac7899…:27o @ client.js?5ac7899…:27g.(anonymous function) @ client.js?5ac7899…:28(anonymous function) @ client.js?5ac7899…:93o @ client.js?5ac7899…:61(anonymous function) @ client.js?5ac7899…:61e.(anonymous function) @ client.js?5ac7899…:61r @ client.js?5ac7899…:2(anonymous function) @ client.js?5ac7899…:2t @ client.js?5ac7899…:1(anonymous function) @ client.js?5ac7899…:2e @ client.js?5ac7899…:93value @ client.js?5ac7899…:93notifyAll @ client.js?5ac7899…:30close @ client.js?5ac7899…:90closeAll @ client.js?5ac7899…:28perform @ client.js?5ac7899…:28perform @ client.js?5ac7899…:28perform @ client.js?5ac7899…:1E @ client.js?5ac7899…:1closeAll @ client.js?5ac7899…:28perform @ client.js?5ac7899…:28batchedUpdates @ client.js?5ac7899…:59i @ client.js?5ac7899…:1dispatchEvent @ client.js?5ac7899…:90
client.js?5ac7899…:60 Uncaught (in promise) TypeError: Cannot read property '_id' of undefined(…)(anonymous function) @ client.js?5ac7899…:60(anonymous function) @ client.js?5ac7899…:2(anonymous function) @ client.js?5ac7899…:91t @ client.js?5ac7899…:91u @ client.js?5ac7899…:60(anonymous function) @ client.js?5ac7899…:93(anonymous function) @ client.js?5ac7899…:88(anonymous function) @ client.js?5ac7899…:91dispatch @ client.js?5ac7899…:91(anonymous function) @ client.js?5ac7899…:93o @ client.js?5ac7899…:61(anonymous function) @ client.js?5ac7899…:61e.(anonymous function) @ client.js?5ac7899…:61r @ client.js?5ac7899…:2(anonymous function) @ client.js?5ac7899…:2
client.js?5ac7899…:61 XHR finished loading: GET "http://[redacted].net/reporting/odata/settings".f.end @ client.js?5ac7899…:61(anonymous function) @ client.js?5ac7899…:28n._execute @ client.js?5ac7899…:26o._resolveFromExecutor @ client.js?5ac7899…:27o @ client.js?5ac7899…:27g.(anonymous function) @ client.js?5ac7899…:28(anonymous function) @ client.js?5ac7899…:93o @ client.js?5ac7899…:61(anonymous function) @ client.js?5ac7899…:61e.(anonymous function) @ client.js?5ac7899…:61r @ client.js?5ac7899…:2(anonymous function) @ client.js?5ac7899…:2t @ client.js?5ac7899…:1(anonymous function) @ client.js?5ac7899…:2(anonymous function) @ client.js?5ac7899…:93(anonymous function) @ client.js?5ac7899…:91(anonymous function) @ client.js?5ac7899…:91(anonymous function) @ client.js?5ac7899…:93o @ client.js?5ac7899…:61(anonymous function) @ client.js?5ac7899…:61e.(anonymous function) @ client.js?5ac7899…:61r @ client.js?5ac7899…:2(anonymous function) @ client.js?5ac7899…:2
client.js?5ac7899…:61 XHR finished loading: GET "http://[redacted].net/reporting/odata/templates?$top=5&$select=name,recipe,modificationDate&$orderby=modificationDate%20desc".f.end @ client.js?5ac7899…:61(anonymous function) @ client.js?5ac7899…:28n._execute @ client.js?5ac7899…:26o._resolveFromExecutor @ client.js?5ac7899…:27o @ client.js?5ac7899…:27g.(anonymous function) @ client.js?5ac7899…:28(anonymous function) @ client.js?5ac7899…:93o @ client.js?5ac7899…:61(anonymous function) @ client.js?5ac7899…:61e.(anonymous function) @ client.js?5ac7899…:61r @ client.js?5ac7899…:2(anonymous function) @ client.js?5ac7899…:2t @ client.js?5ac7899…:1(anonymous function) @ client.js?5ac7899…:2e @ client.js?5ac7899…:93value @ client.js?5ac7899…:93notifyAll @ client.js?5ac7899…:30close @ client.js?5ac7899…:90closeAll @ client.js?5ac7899…:28perform @ client.js?5ac7899…:28perform @ client.js?5ac7899…:28perform @ client.js?5ac7899…:1E @ client.js?5ac7899…:1closeAll @ client.js?5ac7899…:28perform @ client.js?5ac7899…:28batchedUpdates @ client.js?5ac7899…:59i @ client.js?5ac7899…:1dispatchEvent @ client.js?5ac7899…:90
client.js?5ac7899…:61 XHR finished loading: GET "http://[redacted]/reporting/odata/settings".
``` 
in the browser console.  No errors show up in the server-side logs.  Just "OData query on templates", "OData query on settings".   I feel like the error is probably related to the fact that this call: 

[redacted].net/reporting/odata/templates(57717d9e387844200057f98a)

is returning only this: 

{"@odata.context":"http://[redacted].net/reporting/odata/$metadata#templates","value":[]}

Any suggestions?

**Comment by pofider** [comment]
It was tested and it works... at least on my pc.

It looks that the request loading the list of templates returns correctly the items, but the detail request doesn't return any item. Currently I have no idea how this could happen.

Do you have anything interesting to add? Like you have some custom extensions, permissions..?

Maybe you can send me your mongo dump to email?

**Comment by jcfergus** [comment]
I'm running inside a larger express App, but I've stripped out the vast majority of the non-jsreport stuff for testing purposes.  The only custom extension is one I cobbled together to allow me to store a blob of parameters alongside scheduled tasks, which is ... fairly trivial and I don't know how it could be interfering in this way.  

I'll zip both it and a mongo dump up and send to you when I get a few minutes.

**Comment by billsinc** [comment]
I'm having the same issue.  I spun up a new, clean instance as well and once I create a template, I see the same issue when I save it a second time.  Now I can't save/access any of them including the sample files.

**Comment by pofider** [comment]
@billsinc Ok, please give me your configuration -> os, node.js, npm versions

**Comment by billsinc** [comment]
64bit Amazon Linux 2016.03 v2.1.3 running Node.js 4.4.6 behind nginx

**Comment by jcfergus** [comment]
I'm on OSX inside Docker (so Debian Jesse), Node.js v5.9.0 behind nginx.

**Comment by pofider** [comment]
Ok, I can replicate the problem now... my fault. I'll try to find the fix ASAP.

**Comment by billsinc** [comment]
Thanks Jan.

**Comment by pofider** [comment]
There were likely some breaking changes in [mongodb](https://github.com/mongodb/node-mongodb-native) driver 2.x.

It helped to me to downgrade to version mongod@2.0.26
This should be the temporary workaround, until I make it working with the latest.

**Comment by jcfergus** [comment]
Thanks to this, I managed to get my stuff working again - I discovered I had another dependency (connect-mongo, required by express-session) that was pulling mongo 2.2.x in.  Once I massaged things to use an older version of connect-mongo which was dependent on mongo 2.1.x, I'm up and running again.  

Now I just have to update my custom extension to work with the new UI.  *sigh*

**Comment by pofider** [comment]
I updated mongodb driver in jsreport-mongodb-store to the latest and it seems to work fine with several mongodb server versions.

This fix is in jsreport-mongodb-store@0.4.0

**Comment by billsinc** [comment]
Works, thanks.
